cpp/modmesh/linalg/kalman_filter.hpp

```diff
@@ -0,0 +1,523 @@
+    }
+    if (u.ndim() != 1 || u.shape(0) != m_control_size)
+    {
+        Formatter oss;
+        oss << "KalmanFilterImpl::check_control: The control SimpleArray u must be 1D of length control_size (" << m_control_size << "), but got shape (";
+        for (ssize_t i = 0; i < u.ndim(); ++i)
+        {
+            if (i > 0)
+            {
+                oss << ", ";
+            }
+            oss << u.shape(i);
+        }
+        oss << ")";
+        throw std::invalid_argument(oss.str());
+    }
+}
+
+template <typename T>
+auto KalmanFilterImpl<T>::innovation(array_type const & z) -> array_type
+{
+    // y ← z - H·x  (y ← z - m_h·m_x)
+    array_type x_col = m_x.reshape(small_vector<size_t>{m_state_size, 1});
+    array_type hx = m_h.matmul(x_col);
+    array_type hx_vec = hx.reshape(small_vector<size_t>{m_measurement_size});
+    return z.sub(hx_vec);
+}
+
+template <typename T>
+auto KalmanFilterImpl<T>::innovation_covariance() -> array_type
+{
+    // S ← H·P·H^H + R + jitter·I  (s ← m_h·m_p·m_h^H + m_r + m_jitter·I)
+    array_type h_h = hermitian(m_h);
+    array_type hph_h = m_h.matmul(m_p).matmul(h_h);
+    array_type s = hph_h.add(m_r).add(create_noise_matrix(m_measurement_size, m_jitter));
+
+    // S ← 0.5(S + S^H)  (s ← 0.5(s + s^H))
+    return symmetrize(s);
+}
+
+template <typename T>
+auto KalmanFilterImpl<T>::kalman_gain(array_type const & s) -> array_type
+{
+    // K ← P·H^H·S^{-1} via LLT solve
+    array_type B = m_h.matmul(m_p); // H·P = B
+    array_type X = llt_solve(s, B); // S·X = B
+    return hermitian(X); // K = X^H
+}
+
+template <typename T>
+void KalmanFilterImpl<T>::update_state(array_type const & k, array_type const & y)
+{
+    // x ← x + K·y  (m_x ← m_x + k·y)
+    array_type y_col = y.reshape(small_vector<size_t>{m_measurement_size, 1});
+    array_type ky = k.matmul(y_col);
+    array_type ky_vec = ky.reshape(small_vector<size_t>{m_state_size});
+    m_x = m_x.add(ky_vec);
+}
+
+template <typename T>
+void KalmanFilterImpl<T>::update_covariance(array_type const & k)
+{
+    // P ← (I-KH)·P·(I-KH)^H + K·R·K^H  (m_p ← (I-k·m_h)·m_p·(I-k·m_h)^H + k·m_r·k^H)
+    array_type kh = k.matmul(m_h);
+    array_type i_minus_kh = m_i.sub(kh);
+    array_type i_minus_kh_h = hermitian(i_minus_kh);
+    array_type k_h = hermitian(k);
+    array_type term1 = i_minus_kh.matmul(m_p).matmul(i_minus_kh_h);
+    array_type term2 = k.matmul(m_r).matmul(k_h);
+    m_p = term1.add(term2);
+
+    // P ← 0.5(P + P^H)  (m_p ← 0.5(m_p + m_p^H))
+    m_p = symmetrize(m_p);
+}
+
+template <typename T>
+auto KalmanFilterImpl<T>::hermitian(array_type const & a) -> array_type
+{
+    array_type out = a;
+    out.transpose();
+    if constexpr (is_complex_v<T>)
+    {
+        for (auto ptr = out.begin(); ptr != out.end(); ++ptr)
+        {
+            *ptr = ptr->conj();
+        }
+    }
+    return out;
+}
+
+} /* end namespace detail */
+
+template <typename T>
+using KalmanFilter = detail::KalmanFilterImpl<T>;
```

Comment:
Move `detail::KalmanFilterImpl` to `KalmanFilter`.  This alias is not needed.  The class does not need to be in the `detail` namespace either.